americangut/util.py

```diff
@@ -525,3 +530,97 @@ def clean_and_reformat_mapping(in_fp, out_fp, body_site_column_name,
     out_fp.write('\n')
 
     return errors
+
+
+def add_alpha_diversity(map_, alphas):
+    """Adds alpha diversity to the metadata
+
+    Parameters
+    ----------
+    map_ : DataFrame
+        The metadata for all samples
+    alphas : dict
+        A dictionary keying the column name to a dataframe of the alpha
+        diversity information loaded from the `collate_alpha.py` alpha
+        diversity files.
+
+    Returns
+    -------
+    alpha_map : DataFrame
+        A pandas data frame with the alpha diveristy map attached
+    """
+
+    alpha_means = pd.DataFrame.from_dict({
+        '%s' % metric: adf.astype(float).mean(1)
+        for metric, adf in alphas.iteritems()
+        })
+
+    metric = sorted(alpha_means.keys())[-1]
+    alpha_map_ = map_.join(alpha_means)
+
+    keep = alpha_map_[metric].apply(lambda x: not pd.isnull(x))
+    alpha_map_ = alpha_map_.loc[keep]
+
+    return alpha_map_
+
+
+def get_single_id_lists(map_, depths):
+    """Identifies a single sample per individual
+
+    Single samples are identified based on host subject id, and then by
+    randomly selecting samples from avaliable samples at the highest
+    rarefaction depth. If an individual has multiple samples, but only one
+    above the rarefaction threshhold, the sample with the higher sequencing
+    depth is selected, and then inherieted across lists. If there are two
+    samples above the same threshhold, the sample is selected randomly.
+
+    Parameters
+    ----------
+    map_: DataFrame
+        A pandas dataframe where the index designates the sample ID, and
+        columns include 'HOST_SUBJECT_ID' for the unique individual identifier,
+        and 'DEPTH', containing the sequencing depth.
+    depths: iterable
+        The rarefaction depths used for analysis. Depths may be an iterable
+        of floats or castable strings.
+
+    Returns
+    -------
+    single_ids : dict
+        A list of sample IDs where each ID represents a single sample for a
+        subject, and the samples exist at the rarefaction depth specified
+        by the key.
+
+    """
+    num_depths = len(depths)
+
+    single_ids = {depth: [] for depth in depths}
+    single_ids['unrare'] = []
+
+    for hsi, subject in map_.groupby('HOST_SUBJECT_ID'):
+        for depth_id, depth in enumerate(depths[::-1]):
```

Comment:
this implies the `depths` are sorted coming in. Can the docstring be updated to reflect this, or can an explicit sort be used to ensure it?